stgit/commands/refresh.py

```diff
@@ -354,10 +354,21 @@ def func(parser, options, args):
                 'To force a full refresh use --force.'
             )
 
-    # Commit index to temp patch, and absorb it into the target patch.
-    retval, temp_name = make_temp_patch(
-        stack, patch_name, paths, temp_index=path_limiting
-    )
+    # Update index and write tree
+    tree = write_tree(stack, paths, temp_index=path_limiting)
+
+    # Run pre-commit hook, if fails, abort refresh
+    if not options.no_verify:
+        pcval = run_pre_commit_hook(stack.repository)
+        if pcval != utils.STGIT_SUCCESS:
+            raise CmdException(
+                'pre-commit hook failed, review the changes using `stg diff`'
+                ', run `stg add` to add them to index and run `stg refresh` again'
+            )
+
```

Comment:
Another behavior that `git commit` has when running `pre-commit` is that it re-reads the index (discarding any cached version of the index) after the `pre-commit` hook is run. This is done to allow for the `pre-commit` hook to modify the index prior to committing. I believe the same need exists in StGit where we would want to allow `pre-commit` to modify files and update the index prior to refresh.

A potential test for this would be to use a `pre-commit` hook that strips trailing whitespace and then does `git add` to update the index with the strip changes, and then returns 0 (success) such that `stg refresh` would capture the stripped whitespace changes.